common/djangoapps/third_party_auth/lti.py

```diff
@@ -0,0 +1,142 @@
+
+        def add_if_exists(lti_key, details_key):
+            if lti_key in lti_params and lti_params[lti_key]:
+                details.update({details_key: lti_params[lti_key]})
+
+        add_if_exists('email', 'email')
+        add_if_exists('lis_person_name_full', 'fullname')
+        add_if_exists('lis_person_name_given', 'first_name')
+        add_if_exists('lis_person_name_family', 'last_name')
+        return details
+
+    @classmethod
+    def get_validated_lti_params(cls, strategy):
+        request = Request(uri=strategy.request.build_absolute_uri(),
+                          http_method=strategy.request.method, body=strategy.request.body)
+        lti_consumer_key = request.oauth_consumer_key
+        (lti_consumer_valid, lti_consumer_secret, lti_max_timestamp_age) = cls.load_lti_consumer(lti_consumer_key)
+        current_time = calendar.timegm(time.gmtime())
+        return cls._get_validated_lti_params_from_values(request=request, current_time=current_time,
+                                                         lti_consumer_valid=lti_consumer_valid,
+                                                         lti_consumer_secret=lti_consumer_secret,
+                                                         lti_max_timestamp_age=lti_max_timestamp_age)
+
+    @classmethod
+    def _get_validated_lti_params_from_values(cls, request, current_time,
+                                              lti_consumer_valid, lti_consumer_secret, lti_max_timestamp_age):
+        # Taking a cue from oauthlib, to avoid leaking information through a timing attack,
+        # we proceed through the entire validation before rejecting any request for any reason.
+        # However, as noted there, the value of doing this is dubious.
+
+        base_uri = normalize_base_string_uri(request.uri)
+        parameters = collect_parameters(uri_query=request.uri_query, body=request.body)
+        parameters_string = normalize_parameters(parameters)
+        base_string = construct_base_string(request.http_method, base_uri, parameters_string)
+
+        computed_signature = sign_hmac_sha1(base_string, unicode(lti_consumer_secret), '')
+        submitted_signature = request.oauth_signature
+
+        data = {k[0]: k[1] for k in parameters}
+
+        def safe_int(value):
+            try:
+                return int(value)
+            except ValueError:
+                return 0
+            except TypeError:
+                return 0
+
+        oauth_timestamp = safe_int(request.oauth_timestamp)
+
+        valid =  lti_consumer_valid
+        valid &= submitted_signature == computed_signature
+        valid &= request.oauth_version == '1.0'
+        valid &= request.oauth_signature_method == 'HMAC-SHA1'
+        valid &= 'user_id' in data  # Not required by LTI but can't log in without one
+        valid &= oauth_timestamp >= current_time - lti_max_timestamp_age
+        valid &= oauth_timestamp <= current_time
```

Comment:
This can't be written as `valid = valid and (submitted_signature == computed_signature)` because that's vulnerable to the timing leak, but I think one could write it as:

```python
valid =  lti_consumer_valid
valid = (submitted_signature == computed_signature) and valid
valid = (request.oauth_version == '1.0') and valid
valid = (request.oauth_signature_method == 'HMAC-SHA1') and valid
valid = ('user_id' in data) and valid  # Not required by LTI but can't log in without one
valid = (oauth_timestamp >= current_time - lti_max_timestamp_age) and valid
valid = (oauth_timestamp <= current_time) and valid
```

or

```python
checks = [  # All of the following must be true and we must always evaluate each one
    lti_consumer_valid,
    submitted_signature == computed_signature,
    request.oauth_version == '1.0',
    request.oauth_signature_method == 'HMAC-SHA1',
    'user_id' in data,  # Not required by LTI but can't log in without one
    oauth_timestamp >= current_time - lti_max_timestamp_age,
    oauth_timestamp <= current_time,
]
valid = sum(1 for check in checks if check) == len(checks)
```

I'm personally OK with either of these or the original code being left as-is, *but* I'd like to see an explanatory comment added to the code. (Otherwise this could be unwittingly weakened in a future code refactor.)